modules/userId.js

```diff
@@ -0,0 +1,424 @@
+        const getIdResult = item.submodule.getId(item.config.params, consentData);
+
+        // If the getId result has a type of function, it is asynchronous and cannot be called until later
+        if (typeof getIdResult === 'function') {
+          item.callback = getIdResult;
+        } else {
+          // A getId result that is not a function is assumed to be valid user id data, which should be saved to users local storage or cookies
+          setStoredValue(item.config.storage, getIdResult, item.config.storage.expires);
+
+          // cache decoded value (this is copied to every adUnit bid)
+          item.idObj = item.submodule.decode(getIdResult);
+        }
+      }
+    } else if (item.config.value) {
+      // cache decoded value (this is copied to every adUnit bid)
+      item.idObj = item.config.value;
+    }
+
+    carry.push(item);
+    return carry;
+  }, []);
+}
+
+/**
+ * list of submodule configurations with valid 'storage' or 'value' obj definitions
+ * * storage config: contains values for storing/retrieving User ID data in browser storage
+ * * value config: object properties that are copied to bids (without saving to storage)
+ * @param {SubmoduleConfig[]} submoduleConfigs
+ * @param {Submodule[]} enabledSubmodules
+ * @returns {SubmoduleConfig[]}
+ */
+export function getValidSubmoduleConfigs(submoduleConfigs, enabledSubmodules) {
+  if (!Array.isArray(submoduleConfigs)) {
+    return [];
+  }
+
+  // list of browser enabled storage types
+  const validStorageTypes = [];
+  if (utils.localStorageIsEnabled()) {
+    validStorageTypes.push(LOCAL_STORAGE);
+  }
+  if (utils.cookiesAreEnabled()) {
+    validStorageTypes.push(COOKIE);
+  }
+
+  return submoduleConfigs.reduce((carry, submoduleConfig) => {
+    // every submodule config obj must contain a valid 'name'
+    if (!submoduleConfig || typeof submoduleConfig.name !== 'string' || !submoduleConfig.name) {
+      return carry;
+    }
+
+    // Validate storage config
+    // contains 'type' and 'name' properties with non-empty string values
+    // 'type' must be a value currently enabled in the browser
+    if (submoduleConfig.storage &&
+      typeof submoduleConfig.storage.type === 'string' && submoduleConfig.storage.type &&
+      typeof submoduleConfig.storage.name === 'string' && submoduleConfig.storage.name &&
+      validStorageTypes.indexOf(submoduleConfig.storage.type) !== -1) {
+      carry.push(submoduleConfig);
+    } else if (submoduleConfig.value !== null && typeof submoduleConfig.value === 'object') {
+      // Validate value config
+      // must be valid object with at least one property
+      carry.push(submoduleConfig);
+    }
+    return carry;
+  }, []);
+}
+
+/**
+ * @param config
+ * @param {Submodule[]} enabledSubmodules
+ */
+export function init (config, enabledSubmodules) {
+  submodules = [];
+  initializedSubmodules = undefined;
+
+  // exit immediately if opt out cookie exists
+  if (utils.getCookie('_pbjs_id_optout')) {
```

Comment:
Can this be made configurable per submodule? For pubcid, this was _pubcid_optout.  This would provide continuity when the pub switches modules.  

Also this is checking for cookies only.  Would it make sense to check local storage as well?

Reply:
I don't think it's necessary to support per-module optout, but I'd be ok with supporting _pubcid_optout for backwards compatibility. @idettman ?

Reply:
Supporting `_pubcid_optout` makes sense, I'll add it now.

If we wanted to support both `cookie` and `html5`(local storage), we could change the config `storage` obj to an array to support checking `cookie` and `html5`. The array order would check the first item and use the second item if the first item failed.

```
storage: [{
  type: "html5",
  name: "unifiedid",
  expires: 120
}, {
  type: "cookie",
  name: "unifiedid",
  expires: 60
}]
```

Reply:
That's a pretty flexible approach.  It probably makes sense to recommend placing `html5` first in the doc.  

Any thoughts on transition from `cookie` to local storage?  Let's say `html5` is enabled, but not `cookie`.  There is an unifiedid in cookie space but not local storage.  Does it make sense to copy the id over?

Reply:
I think we're over-complicating the storage piece for the first release by copying values between cookies and storage.

The only change I would approve at this point is the minimal change @pycnvr suggested to support optout by looking for _pbjs_id_optout or _pubcid_optout as local-storage if storage.type=html5

Other storage changes can be debated in a different PR.

Reply:
Added checks for `_pbjs_id_optout` and `_pubcid_optout ` in local storage for html5 storage type